### PR TITLE
Allow `.` in package names

### LINKED ISSUE #921: Depedency name must be alphanumeric
I'm trying to use the `gtkmm-4.0` in a project but the issue is that the package name (at least in arch linux) has a dot (`.`) in it, so even though pkg-config is able to find it just fine, poac doesn't seem to like that:
image
image
image

How can this be fixed?
This is Poac version `0.9.3` btw

### PR DESCRIPTION
**Pull request by How2PlayGithub**
fixes #921

Changed the code to check if before and after the `.` is a number, fixing package names like `gtkmm-4.0`.

### CODE REVIEW

**Comment by How2PlayGithub** [comment]
This fixes issue https://github.com/poac-dev/poac/issues/921

**Comment by How2PlayGithub** [comment]
What is going on with poac format...

File: src/Manifest.cc [hunk 1]
```diff
@@ -420,23 +420,32 @@ validateDepName(const StringRef name) {
...
+  // Only alphanumeric characters, `-`, `_`, `/` and `.` if surrounded by
+  // numbers are allowed.
   for (const char c : name) {
-    if (!std::isalnum(c) && c != '-' && c != '_' && c != '/') {
-      throw PoacError("dependency name must be alphanumeric, `-`, `_`, or `/`");
-    }
+    if (!std::isalnum(c) && c != '-' && c != '_' && c != '/'
+        && !(
+            c == '.' && std::isdigit(name[&c - &name[0] - 1])
+            && std::isdigit(name)[&c - &name[0] + 1]
```
Comment (ken-matsui) [inline src/Manifest.cc:429]:
```suggestion
            && std::isdigit(static_cast<unsigned char>(name[&c - &name[0] + 1]))
```

File: src/Manifest.cc [hunk 2]
```diff
@@ -420,23 +420,32 @@ validateDepName(const StringRef name) {
     throw PoacError("dependency name must end with an alphanumeric character");
   }
 
-  // Only alphanumeric characters, `-`, `_`, and `/` are allowed.
+  // Only alphanumeric characters, `-`, `_`, `/` and `.` if surrounded by
+  // numbers are allowed.
```
Comment (ken-matsui) [inline src/Manifest.cc:424]: I would check if these are only the characters for this pass and check if dots are only surrounded by numbers in the next pass.

File: src/Manifest.cc [hunk 3]
```diff
@@ -420,23 +420,32 @@ validateDepName(const StringRef name) {
...
   for (const char c : name) {
-    if (!std::isalnum(c) && c != '-' && c != '_' && c != '/') {
-      throw PoacError("dependency name must be alphanumeric, `-`, `_`, or `/`");
-    }
+    if (!std::isalnum(c) && c != '-' && c != '_' && c != '/'
+        && !(
+            c == '.' && std::isdigit(name[&c - &name[0] - 1])
```
Comment (ken-matsui) [inline src/Manifest.cc:428]:
```suggestion
            c == '.' && std::isdigit(static_cast<unsigned char>(name[&c - &name[0] - 1]))
```

File: src/Manifest.cc [hunk 4]
```diff
@@ -420,23 +420,32 @@ validateDepName(const StringRef name) {
...
   }
 
   // Consecutive `-`, `_`, and `/` are not allowed.
+
   for (usize i = 1; i < name.size(); ++i) {
     if (!std::isalnum(name[i]) && name[i] == name[i - 1]) {
       throw PoacError(
           "dependency name must not contain consecutive non-alphanumeric "
           "characters"
+
```
Comment (ken-matsui) [inline src/Manifest.cc:445]: Unnecessary change.

File: src/Manifest.cc [hunk 5]
```diff
@@ -420,23 +420,32 @@ validateDepName(const StringRef name) {
...
       );
     }
   }
-
```
Comment (ken-matsui) [inline src/Manifest.cc:439]: Unnecessary change.

**Review by ken-matsui — CHANGES_REQUESTED** [review]
Thanks for working on this.  I left some comments.  Please make sure to use the same tool versions as what the CIs are using, i.e., clang-tidy-17 & clang-format-17.

**Comment by ken-matsui** [comment]
@How2PlayGithub  Please avoid spamming the CIs here.  You should try compiling your code locally.  Thanks.

**Comment by How2PlayGithub** [comment]
Quick problem: Whenever I change it to what clang-tidy tells me to do, it just tells me what I originally did

```
/home/usr/Documents/poac/poac-out/debug/../../src/Manifest.cc:427:9: error: boolean expression can be simplified by DeMorgan's theorem [readability-simplify-boolean-expr,-warnings-as-errors]
  427 |     if (!(std::isalnum(c) || c == '-' || c == '_' || c == '/'
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         !                 &&   !=     &&   !=     &&   !=
  428 |           || (c == '.' && (i > 0 && !std::isdigit(name[i - 1]))
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |           &&    !=     ||    <=  ||
  429 |               && (i < name.size() - 1 && !std::isdigit(name[i + 1]))))) {
      |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |               ||    >=                || 
```

```
/home/usr/Documents/poac/poac-out/debug/../../src/Manifest.cc:427:9: error: boolean expression can be simplified by DeMorgan's theorem [readability-simplify-boolean-expr,-warnings-as-errors]
  427 |     if (!(std::isalnum(c) && c != '-' && c != '_' && c != '/'
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         !                 ||   ==     ||   ==     ||   ==
  428 |           && (c != '.' || (i <= 0 || !std::isdigit(name[i - 1]))
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |           ||    ==     &&    >    &&
  429 |               || (i >= name.size() - 1 || !std::isdigit(name[i + 1]))))) {
      |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |               &&    <                  && 
```

**Comment by ken-matsui** [comment]
@How2PlayGithub  I would suggest you work on this first.

> I would check if these are only the characters for this pass and check if dots are only surrounded by numbers in the next pass.

**Comment by How2PlayGithub** [comment]
All done! I've tested it and it works

File: src/Manifest.cc [hunk 6]
```diff
@@ -420,20 +420,33 @@ validateDepName(const StringRef name) {
     throw PoacError("dependency name must end with an alphanumeric character");
```
Comment (ken-matsui) [inline src/Manifest.cc:420]: You should update the above comments as well, i.e., Trailing `-`, `_`, `/`, and `.` are not allowed.

File: src/Manifest.cc [hunk 7]
```diff
@@ -420,20 +420,33 @@ validateDepName(const StringRef name) {
...
   }
 
-  // Only alphanumeric characters, `-`, `_`, and `/` are allowed.
```
Comment (ken-matsui) [inline src/Manifest.cc:423]: This pass should only check if this string only contains alphanumeric chars, `-`, `_`, `/`, and `.`.

File: src/Manifest.cc [hunk 8]
```diff
@@ -420,20 +420,33 @@ validateDepName(const StringRef name) {
...
+            "dependency name must be alphanumeric, '-', '_', '/', or '.' "
+            "surrounded by numbers"
+        );
+      }
+      if (c != '.') {
+        throw PoacError("dependency name must be alphanumeric, '-', '_', or '/'"
+        );
+      }
     }
-  }
```
Comment (ken-matsui) [inline src/Manifest.cc:442]: And then, let's add a new for-loop (that I call pass), which checks if there's a dot NOT surrounded by numbers. It can be after the following check. The following checks comment should be updated.

**Review by ken-matsui — CHANGES_REQUESTED** [review]

File: src/Manifest.cc [hunk 9]
```diff
@@ -415,26 +415,38 @@ validateDepName(const StringRef name) {
...
     );
   }
-  // Trailing `-`, `_`, and `/` are not allowed.
+  // Trailing `-`, `_`, `.`, and `/` are not allowed.
   if (!std::isalnum(name.back())) {
     throw PoacError("dependency name must end with an alphanumeric character");
   }
 
-  // Only alphanumeric characters, `-`, `_`, and `/` are allowed.
+  // Only alphanumeric characters, `-`, `_`, `/` and `.` are allowed
```
Comment (ken-matsui) [inline src/Manifest.cc:423]: What is the rationale for removing the last comma?

Comment (ken-matsui) [inline src/Manifest.cc:423]: Please keep the original comment style. I am using the Oxford comma.

```suggestion
  // Only alphanumeric characters, `-`, `_`, `/`, and `.` are allowed
```

File: src/Manifest.cc [hunk 10]
```diff
@@ -415,26 +415,38 @@ validateDepName(const StringRef name) {
     throw PoacError("dependency name must start with an alphanumeric character"
     );
   }
-  // Trailing `-`, `_`, and `/` are not allowed.
+  // Trailing `-`, `_`, `.`, and `/` are not allowed.
```
Comment (ken-matsui) [inline src/Manifest.cc:418]: Why is the order not consistent with the below comment? Why not this?

```suggestion
  // Trailing `-`, `_`, `/`, and `.` are not allowed.
```

File: src/Manifest.cc [hunk 11]
```diff
@@ -415,26 +415,38 @@ validateDepName(const StringRef name) {
...
     throw PoacError("dependency name must end with an alphanumeric character");
   }
 
-  // Only alphanumeric characters, `-`, `_`, and `/` are allowed.
+  // Only alphanumeric characters, `-`, `_`, `/` and `.` are allowed
   for (const char c : name) {
-    if (!std::isalnum(c) && c != '-' && c != '_' && c != '/') {
-      throw PoacError("dependency name must be alphanumeric, `-`, `_`, or `/`");
+    if (!std::isalnum(c) && c != '-' && c != '_' && c != '/' && c != '.') {
+      throw PoacError("dependency name must be alphanumeric, '-', '_', or '/'");
```
Comment (ken-matsui) [inline src/Manifest.cc:426]: What is the rationale for this change?

File: src/Manifest.cc [hunk 12]
```diff
@@ -415,26 +415,38 @@ validateDepName(const StringRef name) {
...
+            || (i == name.size() - 1 || !std::isdigit(name[i + 1])))) {
       throw PoacError(
-          "dependency name must not contain consecutive non-alphanumeric "
-          "characters"
+          "dependency name must be alphanumeric, '-', '_', '/', or '.' "
+          "surrounded by numbers"
       );
     }
+
+    for (usize i = 1; i < name.size(); ++i) {
```
Comment (ken-matsui) [inline src/Manifest.cc:442]: Why did you remove this comment?

```cc
// Consecutive `-`, `_`, and `/` are not allowed.
```

File: src/Manifest.cc [hunk 13]
```diff
@@ -415,26 +415,38 @@ validateDepName(const StringRef name) {
...
+  for (std::size_t i = 0; i < name.size(); ++i) {
+    const char c = name[i];
+    if (c == '.'
+        && ((i == 0 || !std::isdigit(name[i - 1]))
+            || (i == name.size() - 1 || !std::isdigit(name[i + 1])))) {
       throw PoacError(
-          "dependency name must not contain consecutive non-alphanumeric "
-          "characters"
+          "dependency name must be alphanumeric, '-', '_', '/', or '.' "
+          "surrounded by numbers"
```
Comment (ken-matsui) [inline src/Manifest.cc:438]: Is this pass really checking if there are only them?

File: src/Manifest.cc [hunk 14]
```diff
@@ -415,26 +415,38 @@ validateDepName(const StringRef name) {
...
+    if (!std::isalnum(c) && c != '-' && c != '_' && c != '/' && c != '.') {
+      throw PoacError("dependency name must be alphanumeric, '-', '_', or '/'");
     }
   }
 
-  // Consecutive `-`, `_`, and `/` are not allowed.
-  for (usize i = 1; i < name.size(); ++i) {
-    if (!std::isalnum(name[i]) && name[i] == name[i - 1]) {
+  // Checking if the dot is surrounded by numbers
+  for (std::size_t i = 0; i < name.size(); ++i) {
```
Comment (ken-matsui) [inline src/Manifest.cc:431]: Do we really need to check the first and last element?

File: src/Manifest.cc [hunk 15]
```diff
@@ -415,26 +415,38 @@ validateDepName(const StringRef name) {
...
+    const char c = name[i];
```
Comment (ken-matsui) [inline src/Manifest.cc:432]: Why? `c` seems to be used only once.

File: src/Manifest.cc [hunk 16]
```diff
@@ -415,26 +415,38 @@ validateDepName(const StringRef name) {
...
+    if (c == '.'
+        && ((i == 0 || !std::isdigit(name[i - 1]))
+            || (i == name.size() - 1 || !std::isdigit(name[i + 1])))) {
       throw PoacError(
-          "dependency name must not contain consecutive non-alphanumeric "
-          "characters"
+          "dependency name must be alphanumeric, '-', '_', '/', or '.' "
```
Comment (ken-matsui) [inline src/Manifest.cc:437]: You should follow the current coding style. Please use ` instead of '.

**Review by ken-matsui — CHANGES_REQUESTED** [review]

File: src/Manifest.cc [hunk 17]
```diff
@@ -415,26 +415,37 @@ validateDepName(const StringRef name) {
...
-    if (!std::isalnum(c) && c != '-' && c != '_' && c != '/') {
+    if (!std::isalnum(c) && c != '-' && c != '_' && c != '/' && c != '.') {
       throw PoacError("dependency name must be alphanumeric, `-`, `_`, or `/`");
     }
   }
 
-  // Consecutive `-`, `_`, and `/` are not allowed.
-  for (usize i = 1; i < name.size(); ++i) {
-    if (!std::isalnum(name[i]) && name[i] == name[i - 1]) {
+  // Checking if the dot is surrounded by numbers.
```
Comment (ken-matsui) [inline src/Manifest.cc:430]:
```suggestion
  // `.` must be surrounded by numbers.
```

File: src/Manifest.cc [hunk 18]
```diff
@@ -415,26 +415,37 @@ validateDepName(const StringRef name) {
...
+  for (std::size_t i = 1; i < name.size() - 1; ++i) {
+    if (name[i] == '.'
+        && ((i == 0 || !std::isdigit(name[i - 1]))
+            || (i == name.size() - 1 || !std::isdigit(name[i + 1])))) {
```
Comment (ken-matsui) [inline src/Manifest.cc:434]: We now don't need `i == 0` and `i == name.size() - 1`.

File: src/Manifest.cc [hunk 19]
```diff
@@ -415,26 +415,37 @@ validateDepName(const StringRef name) {
...
       throw PoacError(
-          "dependency name must not contain consecutive non-alphanumeric "
-          "characters"
+          "dependency name must have a `.` surrounded by numbers"
       );
     }
+
+  // Consecutive `-`, `_`, and `/` are not allowed.
```
Comment (ken-matsui) [inline src/Manifest.cc:440]: I think the format is incorrect.

File: src/Manifest.cc [hunk 20]
```diff
@@ -415,26 +415,37 @@ validateDepName(const StringRef name) {
...
+    if (!std::isalnum(c) && c != '-' && c != '_' && c != '/' && c != '.') {
       throw PoacError("dependency name must be alphanumeric, `-`, `_`, or `/`");
     }
   }
 
-  // Consecutive `-`, `_`, and `/` are not allowed.
-  for (usize i = 1; i < name.size(); ++i) {
-    if (!std::isalnum(name[i]) && name[i] == name[i - 1]) {
+  // Checking if the dot is surrounded by numbers.
+  for (std::size_t i = 1; i < name.size() - 1; ++i) {
```
Comment (ken-matsui) [inline src/Manifest.cc:431]:
```suggestion
  for (usize i = 1; i < name.size() - 1; ++i) {
```

**Review by ken-matsui — CHANGES_REQUESTED** [review]

File: src/Manifest.cc [hunk 21]
```diff
@@ -415,25 +415,33 @@ validateDepName(const StringRef name) {
...
   if (!std::isalnum(name.back())) {
     throw PoacError("dependency name must end with an alphanumeric character");
   }
 
-  // Only alphanumeric characters, `-`, `_`, and `/` are allowed.
+  // Only alphanumeric characters, `-`, `_`, `/`, and `.` are allowed.
   for (const char c : name) {
-    if (!std::isalnum(c) && c != '-' && c != '_' && c != '/') {
+    if (!std::isalnum(c) && c != '-' && c != '_' && c != '/' && c != '.') {
       throw PoacError("dependency name must be alphanumeric, `-`, `_`, or `/`");
```
Comment (ken-matsui) [inline src/Manifest.cc:426]:
```suggestion
      throw PoacError("dependency name must be alphanumeric, `-`, `_`, `/`, or `.`");
```

File: src/Manifest.cc (hunk 17 above)
Comment (ken-matsui) [reply src/Manifest.cc:430]: Please do not ignore my review.

File: src/Manifest.cc [hunk 22]
```diff
@@ -415,25 +415,33 @@ validateDepName(const StringRef name) {
...
-  for (usize i = 1; i < name.size(); ++i) {
-    if (!std::isalnum(name[i]) && name[i] == name[i - 1]) {
-      throw PoacError(
-          "dependency name must not contain consecutive non-alphanumeric "
-          "characters"
-      );
+  // Checking if the dot is surrounded by numbers.
+  for (usize i = 1; i < name.size() - 1; ++i) {
+    if (name[i] == '.'
+        && ((!std::isdigit(name[i - 1])) || (!std::isdigit(name[i + 1])))) {
```
Comment (ken-matsui) [inline src/Manifest.cc:433]: Why do you need these parentheses?  Can't we just do this?

```suggestion
        && (!std::isdigit(name[i - 1]) || !std::isdigit(name[i + 1]))) {
```

**Review by ken-matsui — CHANGES_REQUESTED** [review]

File: src/Manifest.cc [hunk 23]
```diff
@@ -415,25 +415,35 @@ validateDepName(const StringRef name) {
...
-      );
+  // `.` must be surrounded by numbers.
+  for (usize i = 1; i < name.size() - 1; ++i) {
+    if (name[i] == '.'
+        && (!std::isdigit(name[i - 1]) || !std::isdigit(name[i + 1]))) {
+      throw PoacError("dependency name must have a `.` surrounded by numbers");
+    }
+
+    // Consecutive `-`, `_`, and `/` are not allowed.
+    for (usize i = 1; i < name.size(); ++i) {
```
Comment (ken-matsui) [inline src/Manifest.cc:440]: Why is this for-loop within the `.` check for-loop?

File: src/Manifest.cc [hunk 24]
```diff
@@ -415,25 +415,35 @@ validateDepName(const StringRef name) {
...
-          "characters"
-      );
+  // `.` must be surrounded by numbers.
+  for (usize i = 1; i < name.size() - 1; ++i) {
+    if (name[i] == '.'
+        && (!std::isdigit(name[i - 1]) || !std::isdigit(name[i + 1]))) {
+      throw PoacError("dependency name must have a `.` surrounded by numbers");
+    }
+
+    // Consecutive `-`, `_`, and `/` are not allowed.
```
Comment (ken-matsui) [inline src/Manifest.cc:439]:
```suggestion
    // Consecutive `-`, `_`, `/`, and `.` are not allowed.
```

File: src/Manifest.cc [hunk 25]
```diff
@@ -415,20 +415,28 @@ validateDepName(const StringRef name) {
...
-  // Consecutive `-`, `_`, and `/` are not allowed.
-  for (usize i = 1; i < name.size(); ++i) {
+  // `.` must be surrounded by numbers.
+  for (usize i = 1; i < name.size() - 1; ++i) {
+    if (name[i] == '.'
+        && (!std::isdigit(name[i - 1]) || !std::isdigit(name[i + 1]))) {
+      throw PoacError("dependency name must have a `.` surrounded by numbers");
+    }
+
+    // Consecutive `-`, `_`, `/`, and `.` are not allowed.
```
Comment (ken-matsui) [inline src/Manifest.cc:439]: Now, this is incorrect.  Previously, we were iterating over `1..name.size()` but now `1..name.size() - 1`.

**Review by ken-matsui — CHANGES_REQUESTED** [review]

**Comment by ken-matsui** [comment]
Hi @How2PlayGithub,

Thank you for your interest in contributing to this project.  After multiple rounds of review, it seems that the required changes are still not correct, and I am concerned that the necessary effort to understand and implement the changes may not have been fully put in.

To maintain the quality of the codebase and respect everyone’s time, I need to set some clear expectations:

1. Please thoroughly review the project code before making changes.
2. Make sure to test your changes thoroughly before submitting them for review.
3. If you are unsure about how to implement a change, feel free to ask for guidance, but make sure to put in the effort to understand the feedback provided.

Given the current situation, I am closing this PR.  I encourage you to take the time to fully understand the changes needed and, once you feel confident, you can submit a new PR.

Thank you for your understanding.

Best regards,
Ken
